feat(pg): support hooking into span when query is started

ethanresnick opened the pull request:
## Which problem is this PR solving?

Right now, it's possible to modify the span generated for a query after results are returned for the query, using the `responseHook`. However, it isn't possible to modify the span at the start of the query, based on the information available when the query is issued (i.e., the args passed to `client.query`), because that information isn't passed to the `responseHook`.

For my use case, I'm using Datadog and am tracing an app that issues relatively few distinct queries. Therefore, it makes sense to identify each query as a distinct Datadog "resource", by adding a `resource.name` attribute to each span, where the value of the attribute is the query text. However, this is currently impossible, because the query text isn't available in the `responseHook`, and there's no other hook where it is available.

## Short description of the changes

- Add a `queryHook` param to support the above use case.

NB: this PR shares the first two commits with #1306, so it may be simpler to review that first. It relies on the code cleanup from #1306 to simplify implementing the `queryHook` here. (The code cleanup now allows the main code path in `_getClientQueryPatch` to have a single, normalized `queryConfig` object that can be passed to the `queryHook`.)

blumamir commented:
Maybe I am missing something, but the query text is currently always captured into `db.statement` span attribute [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/092a2509bcf884e1b997e0eaec3a6ca02cfd2058/plugins/node/opentelemetry-instrumentation-pg/src/utils.ts#L120)
Are you looking to duplicate this attribute in your app and record the same text twice?
I am good with adding this hook if you need it but just wanna make sure it is needed.

ethanresnick commented:
>  Are you looking to duplicate this attribute in your app and record the same text twice?

Yup, exactly. I need to duplicate the value into the `resource.name` attribute to make the Datadog UI work well for my use case (see OP).

blumamir commented:
> > Are you looking to duplicate this attribute in your app and record the same text twice?
> 
> Yup, exactly. I need to duplicate the value into the `resource.name` attribute to make the Datadog UI work well for my use case (see OP).

Are you using OpenTelemetry Collector to send spans to DD? there is the [attributesprocessor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/processor/attributesprocessor) which does just that

If you can't or prefer to use instrumentation hook that is fine and I can review this PR.

ethanresnick commented:
> Are you using OpenTelemetry Collector to send spans to DD?

Unfortunately I'm using the DD agent's built-in support for receiving Otel data ([this approach](https://docs.datadoghq.com/tracing/trace_collection/open_standards/otlp_ingest_in_the_agent/?tab=host)), so that I don't have to run a standalone instance of the collector. I _think_ that the DD agent doesn't support attribute processors (or it doesn't expose that support), so a hook seems easier.

blumamir commented on plugins/node/opentelemetry-instrumentation-pg/src/types.ts:
```diff
@@ -17,6 +17,7 @@
 import type * as pgTypes from 'pg';
 import type * as api from '@opentelemetry/api';
 import { InstrumentationConfig } from '@opentelemetry/instrumentation';
+import { PgClientConnectionParams } from './internal-types';
```
`types.ts` is a public api file for the package, and is not allowed to export internal types. You can read more about the issues it might cause [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/GUIDELINES.md#dependencies)

You can alternatively move the `PgClientConnectionParams` interface from `internal-types.ts` to `types.ts`.

ethanresnick replied:
Great catch. I ended up renaming and adding a comment above the internal type — to make clear how the internal type is different from the related `ConnectionConfig` type that is now exported from @types/pg — and making a separate public type for the hook info, so that it's not coupled to the shape of the internal type (which really is typing an object in the pg library that isn't necessarily even intended to be part of that library's API).

blumamir commented on plugins/node/opentelemetry-instrumentation-pg/src/types.ts:
```diff
@@ -26,12 +27,29 @@ export interface PgInstrumentationExecutionResponseHook {
+   *
+   * @default undefined
+   */
+  queryHook?: PgInstrumentationExecutionQueryHook;
```
Please also update the README.md file with this new config option

blumamir commented on plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:
```diff
@@ -173,101 +173,136 @@ export class PgInstrumentation extends InstrumentationBase {
+            },
+            err => {
+              if (err) {
+                diag.error('Error running query hook', err);
```
nit: we prefer to use `plugin._diag.error` for new code, as it is a component logger that also prepends the package name for each log record.

blumamir commented on plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:
```diff
@@ -173,101 +173,136 @@ export class PgInstrumentation extends InstrumentationBase {
+                  // errors it throws.
+                  values: queryConfig.values as unknown[],
+                  name: queryConfig.name as string | undefined,
+                },
```
nit: The type of this object in `types.ts` is 
```js
query: pgTypes.QueryConfig;
```

Which resolves to 
```js
export interface QueryConfig<I extends any[] = any[]> {
    name?: string | undefined;
    text: string;
    values?: I | undefined;
    types?: CustomTypesConfig | undefined;
}
```

It also includes `types` attribute which I don't know what it is used for, but the instrumentation will not pass it to the hook, where a user might expect that it will.

ethanresnick replied:
Makes sense. I replaced this with a distinct type that explicitly reflects what's passed to the hook.

blumamir commented on plugins/node/opentelemetry-instrumentation-pg/src/instrumentation.ts:
```diff
@@ -173,101 +173,136 @@ export class PgInstrumentation extends InstrumentationBase {
 
-        // Perform the original query
-        const result: unknown = original.apply(this, args as any);
+        if (
```
need to add unittests for this hook. both for successful invocation and for the case where it throws. 
You can follow the tests for the `responseHook` instrumentation config option

ethanresnick commented:
@blumamir Ok, I think I addressed all your comments. I also ended up renaming the hook from `queryHook` to `requestHook`, to match the name of the analogous hook on the redis instrumentation (and for parallelism with `responseHook`). Honestly, `requestHook` feels like a bit of a weird name to me, but I figured it was more important to be consistent.

blumamir reviewed (approved):
LGTM 🥇

ethanresnick commented:
@Rauno56 Given your comment in #1306, I guess this is good to merge too (once CI finishes)? Thanks :)
